vyper/context/namespace.py

```diff
@@ -0,0 +1,102 @@
+import sys
+
+from vyper.exceptions import (
+    CompilerPanic,
+    NamespaceCollision,
+    UndeclaredDefinition,
+)
+
+
+class Namespace(dict):
+    """
+    Dictionary subclass that represents the namespace of a contract.
+
+    Attributes
+    ----------
+    _scopes : List[Set]
+        List of sets containing the key names for each scope
+    _has_builtins: bool
+        Boolean indicating if constant environment variables have been added
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._scopes = []
+        self._has_builtins = False
+
+    def __setitem__(self, attr, obj):
+        if attr in self:
+            obj = super().__getitem__(attr)
+            raise NamespaceCollision(f"'{attr}' has already been declared as a {obj}")
+
+        if self._scopes:
+            self._scopes[-1].add(attr)
+        super().__setitem__(attr, obj)
+
+    def __getitem__(self, key):
+        if key in self:
+            return super().__getitem__(key)
+        raise UndeclaredDefinition(f"'{key}' has not been declared")
```

Comment:
I prefer these flipped
```suggestion
        if key not in self:
            raise UndeclaredDefinition(f"'{key}' has not been declared")
        return super().__getitem__(key)
```